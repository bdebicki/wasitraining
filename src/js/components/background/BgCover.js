import BG from './elementHandlers/background';
import LAYOUT from '../../elementHandlers/layout';
import addSvgElement from '../../utils/addSvgElement';

export default class BgCover {
	static renderCover() {
		const rectSettings = {
			classNames: BG.COVER_CLASS,
			properties: [
				{ height: '100%' },
				{ width: '100%' },
			],
		};
		const svgSettings = {
			id: LAYOUT.BG_COVER_ID,
			children: addSvgElement('rect', rectSettings),
		};

		return addSvgElement('svg', svgSettings);
	}

	static render() {
		return BgCover.renderCover();
	}
}
