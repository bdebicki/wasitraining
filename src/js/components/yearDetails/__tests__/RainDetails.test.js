import cleanDOM from '../../../../../tests/utils/cleanDOM';
import rainyEdition from '../../../../../tests/__mocks__/edition-notSort-noMergeArtists.json';
import newRainyEdition from '../../../../../tests/__mocks__/edition-notSort-customLevels.json';
import sunnyEdition from '../../../../../tests/__mocks__/edition-customOrderExceptHeadliners-mainByDaysAndMergeRest.json';
import BG from '../../background/elementHandlers/background';
import RAIN from '../elementHandlers/rain';
import LINK from '../../../elementHandlers/link';
import DIALOGBOX from '../../../utils/elementHandlers/dialogbox';
import RAIN_INFO_MASK_TYPES from '../../../enums/rainInfoMask';
import RainDetails from '../RainDetails';
import BgCover from '../../background/BgCover';

function mockRainDetails(edition) {
	const rainDetails = new RainDetails(edition);

	document.body.appendChild(rainDetails.render());
}

function mockBgMask(edition) {
	const rainDetails = new RainDetails(edition);

	document.body.appendChild(BgCover.render());
	rainDetails.decorateBgCoverByRainMask();
}

function mockEdition(edition) {
	mockRainDetails(edition);
	mockBgMask(edition);
}

function mockRainMaskPosition() {
	document.querySelector(`.${RAIN.INFO_CLASS}`).getBoundingClientRect = jest.fn(() => ({
		left: 100,
		bottom: 100,
	}));
}

function getMaskId() {
	return document.querySelector(`.${BG.COVER_SHAPE_CLASS}`).getAttributeNS(null, 'mask');
}

describe('tests for RainDetails class', () => {
	afterAll(() => cleanDOM());

	describe('render rain details section', () => {
		afterEach(() => cleanDOM());

		it('renders rain details section for rainy edition', () => {
			// having
			const rainDetails = new RainDetails(rainyEdition);

			// then
			expect(rainDetails.render()).toMatchSnapshot();
		});

		it('renders rain details section for sunny edition', () => {
			// having
			const rainDetails = new RainDetails(sunnyEdition);

			// then
			expect(rainDetails.render()).toMatchSnapshot();
		});
	});
	describe('update rain details section', () => {
		afterEach(() => cleanDOM());

		it('uptade rain info', () => {
			// having
			mockRainDetails(rainyEdition);
			const updatedRainDetails = new RainDetails(sunnyEdition);

			// when
			updatedRainDetails.updateRainInfo();

			// then
			expect(document.querySelector(`.${RAIN.INFO_CLASS}`)).toMatchSnapshot();
		});
		it('update rain day details: from rainy to sunny', () => {
			// having
			mockRainDetails(rainyEdition);
			const updatedRainDetails = new RainDetails(sunnyEdition);

			// when
			updatedRainDetails.updateRainDayDetails();

			// then
			expect(document.getElementById(RAIN.DETAILS_ID)).toBeFalsy();
		});
		it('update rain day details: from sunny to rainy', () => {
			// having
			mockRainDetails(sunnyEdition);
			const updatedRainDetails = new RainDetails(rainyEdition);

			// when
			updatedRainDetails.updateRainDayDetails();

			// then
			expect(document.getElementById(RAIN.DETAILS_ID)).toBeTruthy();
		});
		it('update rain day details: from rainy to rainy', () => {
			// having
			mockRainDetails(rainyEdition);
			const updatedRainDetails = new RainDetails(newRainyEdition);

			// when
			updatedRainDetails.updateRainDayDetails();

			// then
			expect(document.getElementById(RAIN.DETAILS_ID)).toMatchSnapshot();
		});
		it('update rain details', () => {
			// having
			mockRainDetails(rainyEdition);
			const updatedRainDetails = new RainDetails(sunnyEdition);

			// when
			updatedRainDetails.updateRainDetails();

			expect(document.querySelector(`.${RAIN.INFO_YES_CLASS}`)).toBeFalsy();
			expect(document.getElementById(RAIN.DETAILS_ID)).toBeFalsy();
		});
	});
	describe('toggle details dialogbox', () => {
		beforeAll(() => mockRainDetails(rainyEdition));

		it('show details dialogbox', () => {
			// when
			document.querySelector(`.${RAIN.HEADER_CLASS} .${LINK.BASIC_CLASS}`).click();

			// then
			expect(document.getElementById(RAIN.DETAILS_ID).classList.contains(DIALOGBOX.VISIBLE_CLASS)).toBeTruthy();
		});
		it('hide details dialogbox', () => {
			// when
			document.querySelector(`.${RAIN.DETAILS_ID} .${DIALOGBOX.CLOSE_CLASS}`).click();

			// then
			expect(document.getElementById(RAIN.DETAILS_ID).classList.contains(DIALOGBOX.VISIBLE_CLASS)).toBeFalsy();
		});
	});
	describe('decorate rain mask', () => {
		afterEach(() => cleanDOM());

		it('decorate bg cover by yes mask', () => {
			// when
			mockEdition(rainyEdition);

			// then
			expect(getMaskId()).toBe(`url(#${RAIN_INFO_MASK_TYPES.TRUE})`);
		});
		it('decorate bg cover by no mask', () => {
			// when
			mockEdition(sunnyEdition);

			// then
			expect(getMaskId()).toBe(`url(#${RAIN_INFO_MASK_TYPES.FALSE})`);
		});
	});
	describe('update rain mask', () => {
		afterEach(() => cleanDOM());

		it('update bg cover mask to yes', () => {
			// having
			mockEdition(sunnyEdition);

			// when
			const newEditionRainDetails = new RainDetails(rainyEdition);
			newEditionRainDetails.updateBgCoverByRainMask();

			// then
			expect(getMaskId()).toBe(`url(#${RAIN_INFO_MASK_TYPES.TRUE})`);
		});
		it('update bg cover mask to no', () => {
			// having
			mockEdition(rainyEdition);

			// when
			const newEditionRainDetails = new RainDetails(sunnyEdition);
			newEditionRainDetails.updateBgCoverByRainMask();

			// then
			expect(getMaskId()).toBe(`url(#${RAIN_INFO_MASK_TYPES.FALSE})`);
		});
		it('do not update mask in case of same mask type', () => {
			// having
			mockEdition(rainyEdition);

			// when
			const newEditionRainDetails = new RainDetails(rainyEdition);
			newEditionRainDetails.updateBgCoverByRainMask();

			// then
			expect(getMaskId()).toBe(`url(#${RAIN_INFO_MASK_TYPES.TRUE})`);
		});
	});
	describe('update mask position', () => {
		it('update mask position', () => {
			// having
			mockEdition(rainyEdition);

			// when
			mockRainMaskPosition();
			RainDetails.updateBgCoverRainMaskPosition();

			// then
			expect(document.querySelector(`.${BG.MASK_SHAPE_YES}`).style.transform).toBe('translate(77px, 77px)');
		});
	});
});
