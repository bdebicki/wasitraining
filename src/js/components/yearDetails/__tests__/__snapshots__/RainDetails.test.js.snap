// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests for RainDetails class render rain details section renders rain details section for rainy edition 1`] = `
<section
  id="rainSection"
>
  <header
    class="rainHeader"
  >
    <h3
      class="rainHeader__headline"
    >
      Rain
    </h3>
    <a
      class="link link--inverted link--sizeS link--hasIcon rainDetails__link"
      href="#rainDetails"
    >
      more
      <svg
        class="icon--plus link__icon"
        height="5px"
        viewBox="0 0 5 5"
        width="5px"
      >
        <rect
          height="1"
          width="5"
          x="0"
          y="2"
        />
        <rect
          height="5"
          width="1"
          x="2"
          y="0"
        />
      </svg>
    </a>
  </header>
  <div
    class="rainInfoPlaceholder rainInfoPlaceholder--yes"
  >
    yes
  </div>
  <div
    class="dialogbox rainDetails dialogbox--stretched"
    id="rainDetails"
  >
    <header
      class="dialogbox__header"
    >
      <h3
        class="dialogbox__headline"
      >
        Rain
      </h3>
      <a
        class="dialogbox__close"
        href="rainDetails"
        title="hide rain details"
      >
        <svg
          class="icon--close link__icon"
          height="10px"
          viewBox="0 0 10 10"
          width="10px"
        >
          <rect
            class="icon--close__line"
            height="2"
            width="12"
            x="-1"
            y="4"
          />
          <rect
            class="icon--close__line"
            height="12"
            width="2"
            x="4"
            y="-1"
          />
        </svg>
      </a>
    </header>
    <ul
      class="rainDetailsList"
    >
      <li
        class="rainDetailsList__item"
      >
        <span
          class="rainDetailsList__day"
        >
          3rd July
        </span>
        <span
          class="rainDetailsList__rain"
        >
          no
        </span>
      </li>
      <li
        class="rainDetailsList__item"
      >
        <span
          class="rainDetailsList__day"
        >
          4th July
        </span>
        <span
          class="rainDetailsList__rain"
        >
          yes
        </span>
      </li>
      <li
        class="rainDetailsList__item"
      >
        <span
          class="rainDetailsList__day"
        >
          5th July
        </span>
        <span
          class="rainDetailsList__rain"
        >
          yes
        </span>
      </li>
      <li
        class="rainDetailsList__item"
      >
        <span
          class="rainDetailsList__day"
        >
          6th July
        </span>
        <span
          class="rainDetailsList__rain"
        >
          no
        </span>
      </li>
    </ul>
  </div>
</section>
`;

exports[`tests for RainDetails class render rain details section renders rain details section for sunny edition 1`] = `
<section
  id="rainSection"
>
  <header
    class="rainHeader"
  >
    <h3
      class="rainHeader__headline"
    >
      Rain
    </h3>
  </header>
  <div
    class="rainInfoPlaceholder rainInfoPlaceholder--no"
  >
    no
  </div>
</section>
`;

exports[`tests for RainDetails class update rain details section update rain day details: from rainy to rainy 1`] = `
<div
  class="dialogbox rainDetails dialogbox--stretched"
  id="rainDetails"
>
  <header
    class="dialogbox__header"
  >
    <h3
      class="dialogbox__headline"
    >
      Rain
    </h3>
    <a
      class="dialogbox__close"
      href="rainDetails"
      title="hide rain details"
    >
      <svg
        class="icon--close link__icon"
        height="10px"
        viewBox="0 0 10 10"
        width="10px"
      >
        <rect
          class="icon--close__line"
          height="2"
          width="12"
          x="-1"
          y="4"
        />
        <rect
          class="icon--close__line"
          height="12"
          width="2"
          x="4"
          y="-1"
        />
      </svg>
    </a>
  </header>
  <ul
    class="rainDetailsList"
  >
    <li
      class="rainDetailsList__item"
    >
      <span
        class="rainDetailsList__day"
      >
        2nd July
      </span>
      <span
        class="rainDetailsList__rain"
      >
        yes
      </span>
    </li>
    <li
      class="rainDetailsList__item"
    >
      <span
        class="rainDetailsList__day"
      >
        3rd July
      </span>
      <span
        class="rainDetailsList__rain"
      >
        no
      </span>
    </li>
  </ul>
</div>
`;

exports[`tests for RainDetails class update rain details section uptade rain info 1`] = `
<div
  class="rainInfoPlaceholder rainInfoPlaceholder--no"
>
  no
</div>
`;
